x/claim/keeper/hooks.go

```diff
@@ -0,0 +1,95 @@
+package keeper
+
+import (
+	"time"
+
+	"github.com/c-osmosis/osmosis/x/claim/types"
+	gammtypes "github.com/c-osmosis/osmosis/x/gamm/types"
+	lockuptypes "github.com/c-osmosis/osmosis/x/lockup/types"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	govtypes "github.com/cosmos/cosmos-sdk/x/gov/types"
+	stakingtypes "github.com/cosmos/cosmos-sdk/x/staking/types"
+)
+
+func (k Keeper) AfterAddLiquidity(ctx sdk.Context, sender sdk.AccAddress) {
+	k.SetUserAction(ctx, sender, types.ActionAddLiquidity)
+}
+
+func (k Keeper) AfterSwap(ctx sdk.Context, sender sdk.AccAddress) {
+	k.SetUserAction(ctx, sender, types.ActionSwap)
+}
+
+func (k Keeper) AfterProposalVote(ctx sdk.Context, proposalID uint64, voterAddr sdk.AccAddress) {
+	k.SetUserAction(ctx, voterAddr, types.ActionVote)
+}
+
+func (k Keeper) AfterDelegationModified(ctx sdk.Context, delAddr sdk.AccAddress, valAddr sdk.ValAddress) {
```

Comment:
Should we instead use `BeforeDelegationCreated` so that they can use the coins that are unlocking as the delegation amount itself?

Reply:
Also related, we need a way for genesis validators to access these tokens during the gentx.

Reply:
Ok

Reply:
DeliverGenTxs will call BeforeDelegationCreated hook, related claim hook would be called without any code.